pkg/controllers/webhookcerts/cert_requester.go

```diff
@@ -0,0 +1,380 @@
+			return pbool(false), nil
+		}
+
+		klog.V(5).Info("the authenticator's certificate is almost expired, will schedule a new CSR creation")
+
+		// delete the cert/key secret and schedule a new sync
+		if err := c.secretClient.Delete(ctx, certSecretName, metav1.DeleteOptions{}); err != nil {
+			return nil, err
+		}
+		return nil, factory.SyntheticRequeueError
+	}
+
+	// either we don't have a cert yet or it's almost expired
+	csrObj, err := c.csrLister.Get(csrName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	if csrObj != nil {
+		if len(csrObj.Status.Conditions) == 0 {
+			return pbool(true), nil
+		}
+
+		for _, cond := range csrObj.Status.Conditions {
+			switch cond.Type {
+			case certapiv1.CertificateDenied, certapiv1.CertificateFailed:
+				if cond.Status == corev1.ConditionTrue {
+					syncCtx.Recorder().Eventf("CSRFailure", "previous CSR got denied or it otherwise failed: {%s - %s: %s}", cond.Type, cond.Reason, cond.Message)
+					klog.Warning("the previous CSR got denied or it otherwise failed: %v", cond)
+					if err := c.resubmitCSR(ctx, key); err != nil {
+						return nil, err
+					}
+					return pbool(true), nil
+				}
+			}
+		}
+
+		for _, cond := range csrObj.Status.Conditions {
+			if cond.Type == certapiv1.CertificateApproved && cond.Status == corev1.ConditionTrue {
+				csrCert := csrObj.Status.Certificate
+				if len(csrCert) == 0 {
+					return pbool(true), nil
+				}
+				// do a live Get, the cache might be stale at this point if we did an Update removing the cert before
+				certSecret, err := c.secretLister.Secrets("openshift-oauth-apiserver").Get(certSecretName)
+				if err != nil {
+					return nil, err
+				}
+				certSecret.Data["tls.crt"] = csrCert
+				_, err = c.secretClient.Update(ctx, certSecret, metav1.UpdateOptions{})
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+
+	}
+
+	if err = c.resubmitCSR(ctx, key); err != nil {
+		return nil, err
+	}
+
+	return pbool(true), nil
+}
+
+func (c *webhookAuthenticatorCertRequester) getKeyCertFromSecret(ctx context.Context) (key *rsa.PrivateKey, cert *x509.Certificate, keyRotated bool, err error) {
+	certSecret, err := c.secretLister.Secrets("openshift-oauth-apiserver").Get(certSecretName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, nil, false, err
+	}
+
+	var keyPEM []byte
+	if certSecret != nil {
+		keyPEM = certSecret.Data["tls.key"]
+	}
+	if len(keyPEM) > 0 {
+		key, err = pemDecodeRSAPrivate(keyPEM)
+		if err != nil {
+			klog.Error("failed to decode the private key used for the openshift-authenticator, new key will be generated: %v", err)
+			// delete the secret and regenerate it in the next step with a new key
+			err = c.secretClient.Delete(ctx, certSecretName, metav1.DeleteOptions{})
+			if err != nil {
+				return nil, nil, false, err
+			}
+		}
+	}
+	if key == nil {
+		key, err = rsa.GenerateKey(rand.Reader, 4096)
```

Comment:
rand is a global var and doesn't need initialization, right?

Reply:
`rand` is just a package in fact, `Reader` reads from system's crypto-safe random resource